### PR TITLE
Feature/failed event

### PR DESCRIPTION
**Pull request by YetAnotherMatt**

### CODE REVIEW

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/CommonEventType.java [hunk 1]
```diff
@@ -3,5 +3,6 @@
 import uk.gov.dhsc.htbhf.logging.EventType;
 
 public enum CommonEventType implements EventType {
-    APPLICATION_STARTED
+    APPLICATION_STARTED,
+    FAILED_EVENT
```
Comment (mikeSmith08) [inline common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/CommonEventType.java]: As this is an EventType, should this just be `FAILED` rather than `FAILED_EVENT`? Similarly with `EventFailedEvent` below.

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * Represents an event that has failed.
+ */
+public class EventFailedEvent extends Event {
+
+    public static final String FAILED_EVENT_KEY = "failedEvent";
+    public static final String FAILURE_DESCRIPTION_KEY = "failureDescription";
+
+    @Builder
+    public EventFailedEvent(String failureDescription, Event failedEvent) {
```
Comment (mikeSmith08) [inline common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java]: It took me a while to figure out why this was taking in an `Event` on construction so it might be worthwhile having a bit of JavaDoc explaining its usage.

Comment (mikeSmith08) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java]: I'd also do a check here (and unit test) to make sure the `failedEvent` isn't null, I can foresee the possibility in a failure scenario where the original event might be null.

File: common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEventTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        EventFailedEvent failedEvent = EventFailedEvent.builder()
+                .failedEvent(event)
+                .failureDescription("myExceptionMessage")
+                .build();
+
+        // then
+        Map<String, Object> metadata = failedEvent.getEventMetadata();
+        assertThat(metadata).containsAllEntriesOf(event.getEventMetadata());
+        assertThat(metadata.get(FAILED_EVENT_KEY)).isEqualTo(event.getEventType());
+        assertThat(metadata.get(FAILURE_DESCRIPTION_KEY)).isEqualTo("myExceptionMessage");
```
Comment (mikeSmith08) [inline common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEventTest.java]: You should also assert that the `EventType` is set to `FAILED_EVENT` and that the timestamp is set.

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        super(CommonEventType.FAILED_EVENT, failedEvent.getTimestamp(), constructMetadata(failureDescription, failedEvent));
+    }
+
+    private static Map<String, Object> constructMetadata(String failureDescription, Event failedEvent) {
+        Map<String, Object> metadata = new HashMap<>(failedEvent.getEventMetadata());
+        metadata.put(FAILED_EVENT_KEY, failedEvent.getEventType());
```
Comment (mikeSmith08) [inline common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java]: Do we also want to store the timestamp from the original event?

File: common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEventTest.java (hunk 3 above)
Comment (YetAnotherMatt) [reply common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEventTest.java]: done

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java (hunk 4 above)
Comment (YetAnotherMatt) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java]: we're using the original timestamp as the timestamp for the failure event

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java (hunk 2 above)
Comment (YetAnotherMatt) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/EventFailedEvent.java]: hopefully javadoc is no longer necessary/

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/CommonEventType.java (hunk 1 above)
Comment (YetAnotherMatt) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/CommonEventType.java]: now FAILURE

**Review by mikeSmith08 — DISMISSED** [review]

**Review by gethyn1 — DISMISSED** [review]

**Review by dwybourn — APPROVED** [review]

**Review by mikeSmith08 — APPROVED** [review]
